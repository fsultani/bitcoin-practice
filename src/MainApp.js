import React, { Component } from 'react';
import styled from 'styled-components';

import Lottie from 'react-lottie';
import * as animationData from './loader.json';
import axios from 'axios';

import InputForm from './InputForm';
import Totals from './Totals';
import RecentTransactions from './RecentTransactions';
import ErrorModal from './ErrorModal';

const Wrapper = styled.div`
  width: 800px;
  margin: 0 auto;
`;

export default class MainApp extends Component {
  constructor() {
    super();
    this.state = {
      loading: false,
      bitcoinAddress: null,
      total_received: null,
      total_sent: null,
      final_balance: null,
      transactions: [],
      errorMessage: 'null',
      isOpen: false
    };
  }

  handleOpen = () => {
    this.setState({ isOpen: true });
  };

  handleClose = () => {
    this.setState({ isOpen: false });
  };

  handleSubmit = event => {
    event.preventDefault();
    this.setState({ loading: true });
    this.getBitcoinData();
  };

  handleChange = address => {
    this.setState({ bitcoinAddress: address });
  };

  handleError(err) {
    if (err.response && err.response.data) {
      this.setState({
        isOpen: true,
        errorMessage: err.resp.data
      });
    } else {
      // If there's a network error, reload the app
      localStorage.setItem('Error', true);
      window.location.reload();
    }
  }

  refetchBitcoinData() {
    // Set refetch interval to 3 minutes to try to avoid a 429 error
    setInterval(() => this.getBitcoinData(), 180000);
  }

  getBitcoinData() {
    axios
      .get(
        `https://blockchain.info/multiaddr?cors=true&active=${
          this.state.bitcoinAddress
        }`
      )
      .then(res => {
        // console.log('res.data.addresses[0]\n', res.data.addresses[0])
        this.setState(
          {
            loading: false,
            total_received: res.data.addresses[0].total_received / 100000000,
            total_sent: res.data.addresses[0].total_sent / 100000000,
            final_balance: res.data.addresses[0].final_balance / 100000000,
            transactions: res.data.txs,
            displayAddress: this.state.bitcoinAddress
          },
          () => this.refetchBitcoinData()
        );
      })
      .catch(err => {
        this.setState(
          {
            loading: false
          },
          () => this.handleError(err)
        );
      });
  }

  componentDidMount() {
    // Check if the app was reloaded due to a network error and display error message
    if (localStorage.getItem('Error')) {
      this.setState({
        isOpen: true,
        errorMessage: 'Something went wrong'
      });
      localStorage.removeItem('Error');
    }
  }

  render() {
    const defaultOptions = {
      loop: true,
      autoplay: true,
      animationData: animationData,
      rendererSettings: {
        preserveAspectRatio: 'xMidYMid slice'
      }
    };

    return (
      <Wrapper>
        {this.state.loading && (
          <Lottie options={defaultOptions} height={400} width={400} />
        )}

        {this.state.errorMessage && (
          <ErrorModal
            isOpen={this.state.isOpen}
            handleClose={this.handleClose}
            errorMessage={this.state.errorMessage}
          />
        )}

        {!this.state.loading && (
          <div>
            <InputForm
              onChange={this.handleChange}
              handleSubmit={this.handleSubmit}
              value={this.state.value}
            />

            <Totals
              totalReceived={this.state.total_received}
              totalSent={this.state.total_sent}
              finalBalance={this.state.final_balance}
            />

            <RecentTransactions
              transactions={this.state.transactions}
              displayAddress={this.state.displayAddress}
            />
          </div>
        )}
      </Wrapper>
    );
  }
}
