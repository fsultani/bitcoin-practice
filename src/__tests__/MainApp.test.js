import React from 'react';
import MainApp from '../MainApp';
import { shallow } from 'enzyme';

import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

it('returns a valid address', () => {
  const wrapper = shallow(<MainApp
    handleChange={jest.fn()}
    handleSubmit={jest.fn()}
  />)
  wrapper.setState({
    loading: false,
    total_received: 1571174.22487431,
    total_sent: 1570207.45500479,
    final_balance: 966.76986952,
    transactions: [
      {
        time: 1535008717,
        result: 39989776,
        balance: 96676986952,
      },
      {
        time: 1535008682,
        result: 1499989776,
        balance: 96636997176,
      },
      {
        time: 1535008455,
        result: 2716966,
        balance: 95137007400,
      },
      {
        time: 1535008077,
        result: -218137,
        balance: 95134290434,
      },
      {
        time: 1535006582,
        result: 14274049,
        balance: 95134508571,
      }
    ],
    displayAddress: '1J37CY8hcdUXQ1KfBhMCsUVafa8XjDsdCn',
  })
  expect(wrapper).toMatchSnapshot();
});

it('returns an invalid address', () => {
  const wrapper = shallow(<MainApp
    handleChange={jest.fn()}
    handleSubmit={jest.fn()}
    handleClose={jest.fn()}
  />)
  wrapper.setState({
    displayAddress: '1J37CY8hcdUXQ1KfBhMCsUVafa8XjDsdCd',
    errorMessage: "Invalid Bitcoin Address. Please try again.",
    isOpen: true,
  })
  expect(wrapper).toMatchSnapshot();
});