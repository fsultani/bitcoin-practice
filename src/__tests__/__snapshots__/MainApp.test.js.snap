// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns a valid address 1`] = `
<styled.div>
  <Component
    errorMessage="null"
    handleClose={[Function]}
    isOpen={false}
  />
  <div>
    <InputForm
      handleSubmit={[Function]}
      onChange={[Function]}
    />
    <Component
      finalBalance={966.76986952}
      totalReceived={1571174.22487431}
      totalSent={1570207.45500479}
    />
    <Component
      displayAddress="1J37CY8hcdUXQ1KfBhMCsUVafa8XjDsdCn"
      transactions={
        Array [
          Object {
            "balance": 96676986952,
            "result": 39989776,
            "time": 1535008717,
          },
          Object {
            "balance": 96636997176,
            "result": 1499989776,
            "time": 1535008682,
          },
          Object {
            "balance": 95137007400,
            "result": 2716966,
            "time": 1535008455,
          },
          Object {
            "balance": 95134290434,
            "result": -218137,
            "time": 1535008077,
          },
          Object {
            "balance": 95134508571,
            "result": 14274049,
            "time": 1535006582,
          },
        ]
      }
    />
  </div>
</styled.div>
`;

exports[`returns an invalid address 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MainApp
    handleChange={[Function]}
    handleClose={[Function]}
    handleSubmit={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        false,
        <Unknown
          errorMessage="Invalid Bitcoin Address. Please try again."
          handleClose={[Function]}
          isOpen={true}
/>,
        <div>
          <InputForm
                    handleSubmit={[Function]}
                    onChange={[Function]}
                    value={undefined}
          />
          <Unknown
                    finalBalance={null}
                    totalReceived={null}
                    totalSent={null}
          />
          <Unknown
                    displayAddress="1J37CY8hcdUXQ1KfBhMCsUVafa8XjDsdCn"
                    transactions={Array []}
          />
</div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "errorMessage": "Invalid Bitcoin Address. Please try again.",
          "handleClose": [Function],
          "isOpen": true,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <InputForm
              handleSubmit={[Function]}
              onChange={[Function]}
              value={undefined}
/>,
            <Unknown
              finalBalance={null}
              totalReceived={null}
              totalSent={null}
/>,
            <Unknown
              displayAddress="1J37CY8hcdUXQ1KfBhMCsUVafa8XjDsdCn"
              transactions={Array []}
/>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "handleSubmit": [Function],
              "onChange": [Function],
              "value": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "finalBalance": null,
              "totalReceived": null,
              "totalSent": null,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "displayAddress": "1J37CY8hcdUXQ1KfBhMCsUVafa8XjDsdCn",
              "transactions": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          false,
          <Unknown
            errorMessage="Invalid Bitcoin Address. Please try again."
            handleClose={[Function]}
            isOpen={true}
/>,
          <div>
            <InputForm
                        handleSubmit={[Function]}
                        onChange={[Function]}
                        value={undefined}
            />
            <Unknown
                        finalBalance={null}
                        totalReceived={null}
                        totalSent={null}
            />
            <Unknown
                        displayAddress="1J37CY8hcdUXQ1KfBhMCsUVafa8XjDsdCn"
                        transactions={Array []}
            />
</div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "errorMessage": "Invalid Bitcoin Address. Please try again.",
            "handleClose": [Function],
            "isOpen": true,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <InputForm
                handleSubmit={[Function]}
                onChange={[Function]}
                value={undefined}
/>,
              <Unknown
                finalBalance={null}
                totalReceived={null}
                totalSent={null}
/>,
              <Unknown
                displayAddress="1J37CY8hcdUXQ1KfBhMCsUVafa8XjDsdCn"
                transactions={Array []}
/>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "handleSubmit": [Function],
                "onChange": [Function],
                "value": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "finalBalance": null,
                "totalReceived": null,
                "totalSent": null,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "displayAddress": "1J37CY8hcdUXQ1KfBhMCsUVafa8XjDsdCn",
                "transactions": Array [],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
